### PR TITLE
fix ray-triangle intersection

### PR DESCRIPTION
**Pull request by alpcihan**

### CODE REVIEW

**Comment by alpcihan** [comment]
[S4 - 16 - Fix triangle-ray intersection](https://trello.com/c/KGfqMrPW/19-s4-16-fix-triangle-ray-intersection)

File: lib/shaders/pathtracer_compute.wgsl [hunk 1]
```diff
@@ -85,190 +84,139 @@ fn traceRay(ray: Ray, seed: u32) -> vec3f {
...
 }
 
-fn hitTriangle(ray: Ray, v0: vec3<f32>, v1: vec3<f32>, v2: vec3<f32>) -> f32 {
-    const EPSILON: f32 = 0.0000001;
-    
-    var vertex0: vec3<f32> = v0;
-    var vertex1: vec3<f32> = v1;
-    var vertex2: vec3<f32> = v2;
+fn intersectTriangles(ray: Ray, bestHit: ptr<function, HitInfo>) {
+    for(var i: u32 = 0; i < 8712; i+=9) { // TODO: pass triangle count as uniform
```
Comment (umuterc) [inline lib/shaders/pathtracer_compute.wgsl]: We can use arrayLength()

**Review by umuterc — CHANGES_REQUESTED** [review]

**Comment by umuterc** [comment]
Is it normal that I am seeing one triangle in the scene ? 
Screenshot 2023-12-21 at 01 56 49

**Comment by alpcihan** [comment]
> Is it normal that I am seeing one triangle in the scene ? Screenshot 2023-12-21 at 01 56 49

@umuterc 
- did you check from the console if the model loaded successfully?
- also a mismatch between the real triangle count and hard-coded triangle count in the for loop might cause an issue. (Still, this might not be the issue. Because, even in that case, I would expect to see more than one triangle).
- if those two are not the issue, could you send me the obj model you tested from a private message? I will check the issue today.

File: lib/shaders/pathtracer_compute.wgsl (hunk 1 above)
Comment (alpcihan) [reply lib/shaders/pathtracer_compute.wgsl]: @umuterc, I did not know that such a method exists. I will use that.

**Comment by alpcihan** [comment]
> Is it normal that I am seeing one triangle in the scene ? Screenshot 2023-12-21 at 01 56 49

@umuterc

- did you check from the console whether the model loaded successfully?

- also, a mismatch between the real triangle count and the hard-coded triangle count in the for loop might be causing an issue. However, this might not be the problem because even in that case, I would expect to see more than one triangle.

- if neither of these is the issue, could you please send me the OBJ model you tested via private message? I will investigate the issue.

File: lib/shaders/pathtracer_compute.wgsl [hunk 2]
```diff
@@ -85,190 +84,139 @@ fn traceRay(ray: Ray, seed: u32) -> vec3f {
...
-    var invDet: f32 = 1.0 / det;
-    var s: vec3<f32> = ray.origin - vertex0;
-    var u: f32 = invDet * dot(s, rayVecXe2);
+        var det: f32 = dot(edge1, rayVecXe2);
 
-    if (u < 0.0 || u > 1.0) {
-        return -1;
-    }
+        if (det > -EPSILON && det < EPSILON) { // This ray is parallel to this triangle.
+            return;
```
Comment (umuterc) [inline lib/shaders/pathtracer_compute.wgsl]: Here algorithm returns and not checking all triangle intersections. Instead it should continue.

**Review by umuterc — CHANGES_REQUESTED** [review]

**Review by umuterc — DISMISSED** [review]

**Review by umuterc — APPROVED** [review]

**Pull request merged**